### ISSUE TITLE
Nearest-neighbor upsampling regression

### ISSUE THREAD

**Issue by qasfb** [pytorch/pytorch#38985]
## 🐛 Bug

When using GPU F.interpolate with scale_factor < 1 (so it's a downsampling), the backprop sends the gradients on the wrong pixel (bottom-right instead of top-left, for scale = 1/integer). Also it doesn't match the CPU interpolation.

It seems to fail only on nearest-neighbor interpolation, not bilinear.

The bug was introduced in this commit from last year:
https://github.com/pytorch/pytorch/commit/c471a63a395991efb963bdb0ad0eccd892be3fac

## To Reproduce

```
import torch
import torch.nn as nn
import torch.nn.functional as F

a_cpu = torch.Tensor(1,1,4,4).uniform_()
a_gpu = a_cpu.cuda()

a_cpu.requires_grad_();
a_gpu.requires_grad_();

scale_factor = 0.25
def upsample(t):
    return F.interpolate(t, scale_factor=scale_factor, mode='nearest')


upsampled_cpu = upsample(a_cpu)
upsampled_gpu = upsample(a_gpu)

grad_cpu = upsampled_cpu.clone().uniform_()
grad_gpu = grad_cpu.cuda()

upsampled_cpu.backward(grad_cpu)
upsampled_gpu.backward(grad_gpu)

input_grad_cpu = a_cpu.grad
input_grad_gpu = a_gpu.grad

print('output match:', upsampled_cpu.allclose(upsampled_gpu.cpu()))
print('gradinput match:', input_grad_cpu.allclose(input_grad_gpu.cpu()))
print('difference:')
print(input_grad_cpu - input_grad_gpu.cpu())
```

Output:
```
output match: True
gradinput match: False
difference:
tensor([[[[ 0.4625,  0.0000,  0.0000,  0.0000],
          [ 0.0000,  0.0000,  0.0000,  0.0000],
          [ 0.0000,  0.0000,  0.0000,  0.0000],
          [ 0.0000,  0.0000,  0.0000, -0.4625]]]])
```

## Workaround for 1/integer scale factors:
```
# out = F.interpolate(input_tensor, scale_factor=(scale, scale))
inv_scale = 1/scale
int_inv_scale = int(inv_scale)
assert(inv_scale == int_inv_scale)
out = input_tensor[:, :, ::int_inv_scale, ::int_inv_scale]
```

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen @ngimel

**Comment by ezyang** [comment]
cc @jjsjann123

**Comment by jjsjann123** [comment]
Thanks for cc'ing me. I'm self assigning the bug.

**Comment by albanD** [comment]
I would be curious to know how this was not caught by the CI.
@jjsjann123 are you planning on looking at it? Or should I?

**Comment by jjsjann123** [comment]
I can look at this later today.

If we are in a hurry, feel free to take it.

**Comment by albanD** [comment]
Not in a hurry, just want to make sure we understand it so that we can make sure this is not happening for other ops.

**Comment by jjsjann123** [comment]
PR issued: #39055 

python tests were busted and never really tested gpu implementation. I can verify the tests failure after test fix; With proper implementation fix, tests are passing now.

**Comment by albanD** [comment]
Corresponding PR is now merged.
